src/hardware/sblaster.cpp

```diff
@@ -1973,38 +1974,72 @@ static uint8_t CTMIXER_Read() {
 	return ret;
 }
 
+// Called by DSP_WRITE_STATUS to check if the write buffer is at capacity.
+static bool write_buffer_at_capacity()
+{
+	// Is the DSP in an abnormal state and therefore we should consider the
+	// buffer at capacity and unable to receive data?
+	if (sb.dsp.state != DSP_S_NORMAL) {
+		return true;
+	}
+	// Report the buffer as having some room every 8th call, as the buffer
+	// will have run down by some amount after sequential calls.
+	if ((++sb.dsp.write_status_counter % 8) == 0) {
+		return false;
+	}
+	// If DMA isn't running then the buffer's definitely not at capacity.
+	if (sb.dma.mode == DSP_DMA_NONE) {
+		return false;
+	}
+	// Finally, the DMA buffer is considered full until it's able to accept
+	// a full (64 KB) write; which is once we've hit the minimum threshold.
+	return (sb.dma.left > sb.dma.min);
+}
+
+// Sound Blaster DSP status byte
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+// Ref: http://archive.gamedev.net/archive/reference/articles/article443.html
+//
+// Read from 02x0Ch (DSP Write Buffer Status): Bit 7, when 0, indicates that the
+// DSP is ready to receive data through the DSP Write Data or Command port
+// (02x0Ch write).
+//
+// Read from 02x0Eh (DSP Data Available Status): Bit 7, when 1, indicates that
+// the DSP has pending data to be read through the DSP Read Data port (02x0Ah
+// read).
+//
+union BufferStatus {
+	uint8_t data = 0;
+	bit_view<7, 1> has_data;
+	// when 1, the buffer has data and is ready to send
+	// when 0, the buffer doesn't have data and is ready to receive
```

Comment:
Can you put the explanation of the bit _above_ it? That's pretty standard I think, and I've been doing it like that for other register bitview definitions.

Reply:
The linked doco says the low 7 bits are reserved. So I suggest doing this to make the definition complete (e.g. i was wondering, are the other bits unmapped because we're not using them? or are they reserved/undefined?)


```cpp
// Unused bits. They appear to be set to 1 when reading the status register.
bit_view<0, 7> reserved;
```

Reply:
:+1: ; we can default to all bits high.